concurrency/LandingTemplateRegistrationAlternate.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) 2014 Amazon.com, Inc.  All rights reserved.
+ *
+ * Owner: search-experience@
+ */
+
+package concurrency;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Teams that want to register their own templates should create an instance
+ * of some class at runtime that:
+ * - sets up the LandingTemplate instance as desired
+ * - calls the static method on this class to register
+ *
+ * @author ludwick
+ */
+public class LandingTemplateRegistration
+{
+    private static final Log LOGGER = LogFactory.getLog(LandingTemplateRegistration.class);
+
+    private static LandingTemplateRegistration INSTANCE = new LandingTemplateRegistration();
+
+    /**
+     * Get the singleton instance of this object.
+     */
+    public static LandingTemplateRegistration instance() 
+    {
+        return INSTANCE;
+    }
+
+    private Map<String, LandingTemplate> templates = new ConcurrentHashMap<>();
+
+    /**
+     * Register a new template by name and json specification. This method
+     * intentionally swallows errors (with a log) on the assumption that users
+     * of this feature will have tests confirming that their registration took
+     * effect.
+     *
+     * @param name the name for the template (i.e. search-display-style)
+     * @param jsonSpecification a string blob of json with fields from {@link LandingTemplate}
+     */
+    public void register(String name, String jsonSpecification)
+    {
+        // __ADI__ I have this LandingTemplateRegistration to show that, if the throw statement does NOT
+        //      post-dominate its branch, the messaging will not mention it
+        //  Note that, by how I shaped the outer IF, templates.containsKey(name) still controls the templates.put(name, template);
+        if (templates.containsKey(name))
+        {
+            if(name.startsWith("lala")) {
+                throw new IllegalStateException("Attempt to register a template twice for name '"+name+"'");
+            }
+        } else {
+
+            try
+            {
+                LandingTemplate template = LandingTemplate.fromJson(jsonSpecification);
+                templates.put(name, template);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **55 and 65**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).